posts/2018-12-21-Procedural-Macros-in-Rust-2018.md

```diff
@@ -0,0 +1,363 @@
+stabilization. It would be infeasible to stabilize the Rust AST because that
+means we could never change it. (imagine if we couldn't have added the `?`
+operator!)
+
+By using `TokenStream` to communicate with procedural macros the compiler is
+able to add new syntactical language features while also being able to compile
+and work with older procedural macros. Let's see now, though, how we can
+actually get useful information out of a `TokenStream`.
+
+### Parsing a `TokenStream`
+
+If `TokenStream` is just a simple iterator, then we've got a long way to go from
+that to an actual parsed function. Although the code is already lexed for us
+we still need to write a whole Rust parser! Thankfully though the community has
+been hard at work to make sure writing procedural macros in Rust is as smooth as
+can be, so you need look no further than the [`syn` crate][syn].
+
+With the [`syn`][syn] crate we can parse any Rust AST as a one-liner:
+
+```rust
+#[proc_macro_attribute]
+pub fn hello(attr: TokenStream, item: TokenStream) -> TokenStream {
+    let input = syn::parse_macro_input!(item as syn::ItemFn);
+    let name = &input.ident;
+    let abi = &input.abi;
+    // ...
+}
+```
+
+The [`syn`][syn] crate not only comes with the ability to parse built-in syntax
+but you can also easily write a recursive descent parser for your own syntax.
+The [`syn::parse` module][spm] has more information about this capability.
+
+### Producing a `TokenStream`
+
+Not only do we take a `TokenStream` as input with a procedural macro, but we
+also need to produce a `TokenStream` as output. This output is typically
+required to be valid Rust syntax, but like the input it's just list of tokens
+that we need to build somehow.
+
+Technically the only way to create a `TokenStream` is via its `FromIterator`
+implementation, which means we'd have to create each token one-by-one and
+collect it into a `TokenStream`. This is quite tedious, though, so let's take a
+look at [`syn`][syn]'s sibling crate: [`quote`].
+
+The [`quote`] crate is a quasi-quoting implementation for Rust which primarily
+provides a convenient macro for us to use:
+
+```rust
+use quote::quote;
+
+#[proc_macro_attribute]
+pub fn hello(attr: TokenStream, item: TokenStream) -> TokenStream {
+    let input = syn::parse_macro_input!(item as syn::ItemFn);
+    let name = &input.ident;
+
+    // Our input function is always equivalent to returning 42, right?
+    let result = quote! {
+        fn #name() -> u32 { 42 }
+    };
+    result.into()
+}
+```
+
+The [`quote!` macro] allows you to write mostly-Rust syntax and interpolate
+variables quickly from the environment with `#foo`. This removes much of the
+tedium of creating a `TokenStream` token-by-token and allows quickly cobbling
+together various pieces of syntax into one return value.
+
+### Tokens and `Span`
+
+Perhaps the greatest feature of procedural macros in Rust 2018 is the ability to
```

Comment:
I would lead with what spans let you do (give good error messages) and come to what spans actually are second.